pmm/summary.go

```diff
@@ -0,0 +1,270 @@
+/*
+	Copyright (c) 2016, Percona LLC and/or its affiliates. All rights reserved.
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <http://www.gnu.org/licenses/>
+*/
+
+package pmm
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/percona/kardianos-service"
+)
+
+// Collector parameters and description.
+type Collector struct {
+	CollectorDescription string
+	ExecCommand          []string
+	OutputFileName       string
+}
+
+// CollectData runs a command and collects output into a file.
+func (c *Collector) CollectData() error {
+	fmt.Printf("%s ... ", c.CollectorDescription)
+	dstfNetworks, err := os.OpenFile(c.OutputFileName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
+	if err != nil {
+		fmt.Printf("Skipped. Failed to create file %s with %s\n", c.OutputFileName, err)
+		return err
+	}
+	defer dstfNetworks.Close()
+	cmd, err := exec.Command(c.ExecCommand[0], c.ExecCommand[1:]...).Output()
+	if err != nil {
+		fmt.Printf("Skipped. Error collecting data by %s: %s\n", c.ExecCommand, err)
+		return err
+	}
+	if _, err := dstfNetworks.Write(cmd); err != nil {
+		fmt.Printf("Skipped. Failed to copy %s output into file %s\n", c.ExecCommand[0], c.OutputFileName)
+		return err
+	}
+	fmt.Println("Done.")
+
+	return err
+}
+
+// CheckMonitoredDBServices finds out what DB instances are monitored.
+func CheckMonitoredDBServices() []string {
+	var monitoredDBServices []string
+	cmdPmmList, err := exec.Command("pmm-admin", "list").Output()
+	if err != nil {
+		fmt.Println("Error exec pmm-admin list", err)
+		return nil
+	}
+	if strings.Contains(string(cmdPmmList), "mysql:metrics") {
+		monitoredDBServices = append(monitoredDBServices, "mysql")
+	}
+	if strings.Contains(string(cmdPmmList), "mongodb:metrics") {
+		monitoredDBServices = append(monitoredDBServices, "mongodb")
+	}
+
+	return monitoredDBServices
+}
+
+// Copy a file for collecting into the final archive
+func copyFile(dirname, name string) error {
+	srcf, err := os.Open(name)
+	if err != nil {
+		return fmt.Errorf("copy %s to %s: %v", name, dirname, err)
+	}
+	defer srcf.Close()
+
+	dstf, err := os.OpenFile(filepath.Join(dirname, filepath.Base(name)), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
+	if err != nil {
+		return fmt.Errorf("copy %s to %s: %v", name, dirname, err)
+	}
+
+	if _, err := io.Copy(dstf, srcf); err != nil {
+		dstf.Close()
+		defer dstf.Close()
```

Comment:
Remove that line – you are closing dstf twice.

Reply:
done